src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -415,10 +418,22 @@ internal SynthesizedAttributeData SynthesizeDebuggerStepThroughAttribute()
             return TrySynthesizeAttribute(WellKnownMember.System_Diagnostics_DebuggerStepThroughAttribute__ctor);
         }
 
-        internal SynthesizedAttributeData SynthesizeReadOnlyAttribute()
+        internal void EnsureIsReadOnlyAttributeExists(Symbol readOnlySymbol)
         {
-            // PROTOTYPE(readonlyRefs) it is optional now as it will be generated in the next PR
-            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_ReadOnlyAttribute__ctor, isOptionalUse: true);
+            // This will report a DeclarationDiagnostics error if the member is not found.
+            // We only need to report this when compiling NetModules, as for other compilations, we will generate it.
+
+            var attributeSymbol = Binder.GetWellKnownTypeMember(
+                compilation: this,
+                member: WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor,
+                diagnostics: DeclarationDiagnostics,
+                location: readOnlySymbol.Locations.FirstOrDefault() ?? Location.None,
+                isOptional: Options.OutputKind != OutputKind.NetModule);
```

Comment:
>isOptional: Options.OutputKind != OutputKind.NetModule [](start = 16, length = 54)

It feels like this doesn't give us the behavior we want. I think we want to embed the attribute only if the type cannot be found, the lack of the required constructor should result in an error rather than in an embedded attribute.